userspace/libscap/engine/bpf/scap_bpf.c

```diff
@@ -1580,6 +1604,127 @@ int32_t scap_bpf_get_stats(struct scap_engine_handle engine, OUT scap_stats* sta
+			{
+				return scap_errprintf(handle->m_lasterr, -ret, "Error looking up local state %d", cpu);
+			}
+			stats[N_EVTS].value.u64 += v.n_evts;
+			stats[N_DROPS_BUFFER_TOTAL].value.u64 += v.n_drops_buffer;
+			stats[N_DROPS_BUFFER_CLONE_FORK_ENTER].value.u64 += v.n_drops_buffer_clone_fork_enter;
+			stats[N_DROPS_BUFFER_CLONE_FORK_EXIT].value.u64 += v.n_drops_buffer_clone_fork_exit;
+			stats[N_DROPS_BUFFER_EXECVE_ENTER].value.u64 += v.n_drops_buffer_execve_enter;
+			stats[N_DROPS_BUFFER_EXECVE_EXIT].value.u64 += v.n_drops_buffer_execve_exit;
+			stats[N_DROPS_BUFFER_CONNECT_ENTER].value.u64 += v.n_drops_buffer_connect_enter;
+			stats[N_DROPS_BUFFER_CONNECT_EXIT].value.u64 += v.n_drops_buffer_connect_exit;
+			stats[N_DROPS_BUFFER_OPEN_ENTER].value.u64 += v.n_drops_buffer_open_enter;
+			stats[N_DROPS_BUFFER_OPEN_EXIT].value.u64 += v.n_drops_buffer_open_exit;
+			stats[N_DROPS_BUFFER_DIR_FILE_ENTER].value.u64 += v.n_drops_buffer_dir_file_enter;
+			stats[N_DROPS_BUFFER_DIR_FILE_EXIT].value.u64 += v.n_drops_buffer_dir_file_exit;
+			stats[N_DROPS_BUFFER_OTHER_INTEREST_ENTER].value.u64 += v.n_drops_buffer_other_interest_enter;
+			stats[N_DROPS_BUFFER_OTHER_INTEREST_EXIT].value.u64 += v.n_drops_buffer_other_interest_exit;
+			stats[N_DROPS_SCRATCH_MAP].value.u64 += v.n_drops_scratch_map;
+			stats[N_DROPS_PAGE_FAULTS].value.u64 += v.n_drops_pf;
+			stats[N_DROPS_BUG].value.u64 += v.n_drops_bug;
+			stats[N_DROPS].value.u64 += v.n_drops_buffer + \
+				v.n_drops_scratch_map + \
+				v.n_drops_pf + \
+				v.n_drops_bug;
+		}
+		i = MAX_KERNEL_COUNTERS_STATS;
+	}
+
+	/* LIBBPF STATS */
+
+	/* At the time of writing (Apr 2, 2023) libbpf stats are only available on a per program granularity.
+	 * This means we cannot measure the statistics for each filler / tail call individually.
+	 * Hopefully someone upstreams such capabilities to libbpf one day :)
+	 * Meanwhile, we can simulate perf comparisons between future LSM hooks and sys enter and exit tracepoints
+	 * via leveraging syscall selection mechanisms `handle->curr_sc_set`.
+	 */
+	if ((flags & PPM_SCAP_STATS_LIBBPF_STATS))
+	{
+		for(int bpf_prog = 0; bpf_prog < BPF_PROG_ATTACHED_MAX; bpf_prog++)
+		{
+			fd = handle->m_attached_progs[bpf_prog].fd;
+			if (fd > 0)
+			{
+				struct bpf_prog_info info = {};
+				__u32 len = sizeof(info);
+				if((ret = bpf_obj_get_info_by_fd(fd, &info, &len)))
+				{
+					// todo: possibly add a warning message or just ignore?
+				}
+				else
+				{
+					for(int stat =  0;  stat < MAX_LIBBPF_STATS; stat++)
+					{
+						if (i > *nstats - 1)
+						{
+							return SCAP_FAILURE;
+						}
+
+						stats[i].valid = true;
+						stats[i].flags = 0;
+						stats[i].flags |= PPM_SCAP_STATS_LIBBPF_STATS;
+						strlcpy(stats[i].name, info.name, STATS_NAME_MAX);
+						if (stat == RUN_CNT)
+						{
+							strncat(stats[i].name, libbpf_stats_names[RUN_CNT], strlen(libbpf_stats_names[RUN_CNT]));
```

Comment:
@LucaGuerra not too happy about `strncat` as it gives warning around "specified bound equals source length", we could keep it for now, but given we invest in cleaning up and adopt best secure coding practices we could look into adding `strlcat` and perhaps we could do another audit in this regard in general? More recommendations? strings always make me nervous.

Reply:
The length parameter should contain the free space in the destination, not the length of the source :)

Still, since it requires you to do the calculation manually, it's not much of an improvement over strlcpy:

```
char dest[100] = "foo";
size_t dest_len = strlen(dest);

// this seems to be the right way to use strncpy:
strncpy(dest, src, sizeof(dest) - dest_len);

// seems equivalent to:
strlcpy(dest + dest_len, src, sizeof(dest) - dest_len);
```

So yeah, we can consider a strlcat that hides the strlen call inside

Reply:
jeez, thanks for catching this @gnosek while I looked again also saw I messed up 2 returns, fixed them in same new commit.